articles/kaneshin.re

```diff
@@ -0,0 +1,414 @@
+    0x0021 00033 (main.go:18)       FUNCDATA        $3, "".init.1.stkobj(SB)
+//}
+
+== パッケージの初期化
+
+Goの作法ではパッケージの初期化と、パッケージ間の依存関係を意識することになります。実行時にパッケージの変数を初期化することによって変数の確保や処理時間を効率化できます。
+パッケージの初期化が多いと@<code>{main}関数が実行されるまで待ち時間やメモリのスコープが関数ではなくパッケージスコープとなることによって変数を参照のコストが高くなることがあります。
+このようなデメリットを理解したうえで初期化をすることは効率性を向上させるために重要です。
+
+=== 初期化の順序
+
+@<code>{init}関数の実行順序は上から順に実行されますが、複数のファイルやパッケージ変数、依存パッケージが存在するときの順序について知らない人もいるのではないでしょうか。パッケージの初期化はパッケージ変数の初期化の後に@<code>{init}関数の実行という順序で行われます。
+
+//list[initialize_package_main][main.go][go]{
+package main
+
+import "runtime"
+
+var _ = submain()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.main.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func submain() int {
+  runtime.Callers(1, pcs[:])
+  logger.a.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+
+func main() {
+  runtime.Callers(1, pcs[:])
+  logger.main.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+//}
+
+//list[initialize_package_a][a.go][go]{
+package main
+
+import "runtime"
+
+var _ = suba()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.a.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func suba() int {
+  runtime.Callers(1, pcs[:])
+  logger.a.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+//}
+
+//list[initialize_package_z][z.go][go]{
+package main
+
+import "runtime"
+
+var _ = subz()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.z.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func subz() int {
+  runtime.Callers(1, pcs[:])
+  logger.z.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+//}
+
+ファイル名は、ファイル名の昇順でわかりやすいものを定義しているとの、出力と呼び出された関数オブジェクトの名前を取得するために次の変数を定義しています。
+
+//list[initialize_package_logger][][go]{
+var logger = struct {
+  a    *log.Logger
+  z    *log.Logger
+  main *log.Logger
+}{
+  a:    log.New(os.Stdout, "   a.go: ", 0),
+  z:    log.New(os.Stdout, "   z.go: ", 0),
+  main: log.New(os.Stdout, "main.go: ", 0),
+}
+
+var pcs [1]uintptr
+//}
+
+//cmd{
+$ go build -o main . && ./main
+   a.go: main.suba
+   a.go: main.submain
+   z.go: main.subz
+   a.go: main.init.0
+main.go: main.init.1
+   z.go: main.init.2
+main.go: main.main
+//}
+
+それぞれのファイルにある変数の初期化が処理され、その後に@<code>{init}関数が処理されています。@<code>{main}関数のあるファイルが最後に初期化処理されるわけではありませんし、main.goやinit.goというファイル名を作ってもファイル名による優先処理はありません。
+
+==== 依存パッケージの初期化
+
+あるパッケージに別のパッケージが依存している場合、依存パッケージが先に初期化処理をするように実行されます。この初期化は再帰的に行われるため、依存の依存パッケージが存在すれば先に依存の依存パッケージの初期化処理がされます。
+
+//list[initialize_package_dep_a][dep/a.go][go]{
+package dep
+
+import "log"
+import "os"
+import "runtime"
+
+var logger = log.New(os.Stdout, " d/a.go: ", 0)
+
+var pcs [1]uintptr
+
+var _ = suba()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func suba() int {
+  runtime.Callers(1, pcs[:])
+  logger.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+//}
+
+@<code>{import _ "github.com/kaneshin/foo/bar/dep"}のようにブランクインポートさせると次のように先に依存パッケージの初期化処理が行われて、依存元のパッケージの初期化が処理されます。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 238 sentence length(106) exceeds the maximum sentence length of 100.
Over 6 characters. (ja-technical-writing/sentence-length)